src/heartratetask/HeartRateTask.cpp

```diff
@@ -25,77 +27,57 @@ void HeartRateTask::Process(void* instance) {
 
 void HeartRateTask::Work() {
   int lastBpm = 0;
+
   while (true) {
+    TickType_t delay = CurrentTaskDelay();
     Messages msg;
-    uint32_t delay;
-    if (state == States::Running) {
-      if (measurementStarted) {
-        delay = ppg.deltaTms;
-      } else {
-        delay = 100;
-      }
-    } else {
-      delay = portMAX_DELAY;
-    }
 
-    if (xQueueReceive(messageQueue, &msg, delay)) {
+    if (xQueueReceive(messageQueue, &msg, delay) == pdTRUE) {
       switch (msg) {
         case Messages::GoToSleep:
-          StopMeasurement();
-          state = States::Idle;
+          if (state == States::Running) {
+            state = States::Idle;
+          } else if (state == States::Measuring) {
+            state = States::BackgroundWaiting;
+            StartWaiting();
+          }
           break;
         case Messages::WakeUp:
-          state = States::Running;
-          if (measurementStarted) {
-            lastBpm = 0;
+          if (state == States::Idle) {
+            state = States::Running;
+          } else if (state == States::BackgroundMeasuring) {
+            state = States::Measuring;
+          } else if (state == States::BackgroundWaiting) {
+            state = States::Measuring;
             StartMeasurement();
           }
           break;
         case Messages::StartMeasurement:
-          if (measurementStarted) {
+          if (state == States::Measuring || state == States::BackgroundMeasuring) {
             break;
           }
+          state = States::Measuring;
           lastBpm = 0;
           StartMeasurement();
-          measurementStarted = true;
           break;
         case Messages::StopMeasurement:
-          if (!measurementStarted) {
+          if (state == States::Running || state == States::Idle) {
             break;
           }
+          if (state == States::Measuring) {
+            state = States::Running;
+          } else if (state == States::BackgroundMeasuring) {
+            state = States::Idle;
+          }
           StopMeasurement();
```

Comment:
I didn't test the branch yet, but I am a bit confused on the `States::BackgroundWaiting` and `States::BackgroundMeasuring`. Why is the latter one handled on `StopMeasurement` but the former one is not?

Also I am trying to find out if this can actually ever be the case, that `Messages::StopMeasurement` comes in and `state == States::BackgroundMeasuring` as to send `Messages::StopMeasurement` it must before getting into `Messages::WakeUp` where both `Background*` states are changed to `States::Measuring`.

Reply:
I think you are correct in your assumption that this branch can't be reached.
The `Messages::StopMeasurement` message is sent when the `Stop` function of the heartrate controller is called. This only happens in the heart rate screen when the start/stop button is pushed. The screen therefore needs to be activated.

The question now is should `BackgroundWaiting` also be handled (in case in the future there might be a background task that stops heart rate measuring and the heartrate task should be able to handle that) or should the `BackgroundMeasuring` branch be removed since right now it is not possible to reach that branch.

Reply:
I am uncertain to be honest. But as I am not sure where/how to document that you should never send `StopMeasurement` from a Background task, I would assume it would be better to handle both `Background*` things in here and set the state to Idle in both cases?

Reply:
I strongly vote for the robust implementation.

Reply:
you should probably handle all cases, maybe using a switch so that the compile fails if you forget one.

Reply:
I agree. State machines based on enums are almost always best implemented by switches.

Reply:
Sounds like a good idea. Will do that!

Reply:
Implemented using switch and handled the `StopMeasurment` case for `Background*`